#### Using the Chicago Crime Dataset(any of the 3 sheets you choose), perfrom the following data preparation steps:
1. Use a funtion to Drop redundant columns (a redundant column is one that gives infromation that another column already explains: eg ID and Case number) 
2. Use functions to create new columns for Months, Day, Season.
3. Use subsetting and grouping to show how the frequency of crime is distributed within Months, Days, Seasons with the most crime record and what crimes are highest and lowest?
4.  According to location description, where does crime happen the most?
5.  Did the types of crime change as years go by? If yes/no. Let the data show the insight.

In [36]:
# Loading necessary libraries
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [2]:
# Saving the file path of the data in a variable 'location'
location = r'C:\Users\Alu\Documents\Altschool Africa\Data'

In [3]:
# Saving the name of the file in a variable 'filename'
filename = '\chicago2.csv'

In [4]:
# Loading the data
chicago_crime_df = pd.read_csv(location + filename)

In [5]:
# Checking the first 5 rows of the dataset
chicago_crime_df.head()

Unnamed: 0       ID Case Number                    Date  \
0       53000  4809324    HM300447  04/19/2006 04:45:02 PM   
1       53001  2599849    HJ197748  02/21/2003 06:30:00 PM   
2       53002  7495181    HR183541  02/22/2009 07:42:00 PM   
3       53003  2628412    HJ231141  03/07/2003 02:46:00 PM   
4       53004  8389615    HT622350  12/06/2011 12:00:00 AM   

                    Block  IUCR   Primary Type  \
0      090XX S NORMAL AVE  1812      NARCOTICS   
1    039XX N AVONDALE AVE  0820          THEFT   
2  011XX S CALIFORNIA AVE  0486        BATTERY   
3      025XX W ADDISON ST  0890          THEFT   
4          072XX S MAY ST  2825  OTHER OFFENSE   

                      Description      Location Description  Arrest  Domestic  \
0  POSS: CANNABIS MORE THAN 30GMS                 RESIDENCE   False     False   
1                  $500 AND UNDER                    STREET   False     False   
2         DOMESTIC BATTERY SIMPLE                     OTHER   False     False   
3                   FROM BUILDING  SCHOOL, PUBLIC, BUILDING   False     False   
4         HARASSMENT BY TELEPHONE                 APARTMENT   False      True   

   Beat  District  Ward  Community Area FBI Code  X Coordinate  Y Coordinate  \
0  2223      22.0  21.0            71.0       18     1174545.0     1845186.0   
1  1731      17.0  38.0            16.0       06     1147850.0     1926182.0   
2  1135      11.0  28.0            29.0      08B           NaN           NaN   
3  1913      19.0  47.0             5.0       06     1158854.0     1923830.0   
4   733       7.0  17.0            68.0       26     1169905.0     1856761.0   

   Year              Updated On   Latitude  Longitude  \
0  2006  02/10/2018 03:50:01 PM  41.730582 -87.636172   
1  2003  02/28/2018 03:56:25 PM  41.953398 -87.731888   
2  2009  08/17/2015 03:03:40 PM        NaN        NaN   
3  2003  02/28/2018 03:56:25 PM  41.946725 -87.691501   
4  2011  02/04/2016 06:33:39 AM  41.762447 -87.652835   

                        Location  
0  (41.730582057, -87.636172403)  
1   (41.953397577, -87.73188822)  
2                            NaN  
3  (41.946724511, -87.691501175)  
4  (41.762447305, -87.652834901)

In [6]:
# Checking the last 5 rows of the dataset
chicago_crime_df.tail()

Unnamed: 0        ID Case Number                    Date  \
995       53995   9619330    HX269158  05/19/2014 06:00:00 PM   
996       53996  11053559    JA390202  08/11/2017 04:00:00 PM   
997       53997  12023154    JD201316  03/21/2020 09:00:00 AM   
998       53998   1604980     G367198  06/23/2001 05:00:00 PM   
999       53999   2660948    HJ275409  04/01/2003 10:00:00 AM   

                   Block  IUCR        Primary Type               Description  \
995   088XX S WABASH AVE  5002       OTHER OFFENSE     OTHER VEHICLE OFFENSE   
996  137XX S JEFFERY AVE  0610            BURGLARY            FORCIBLE ENTRY   
997   058XX N WHIPPLE ST  1130  DECEPTIVE PRACTICE  FRAUD OR CONFIDENCE GAME   
998   070XX S OGLESBY AV  0820               THEFT            $500 AND UNDER   
999    030XX S THROOP ST  0620            BURGLARY            UNLAWFUL ENTRY   

               Location Description  Arrest  Domestic  Beat  District  Ward  \
995                          STREET   False      True   632       6.0   6.0   
996  FACTORY/MANUFACTURING BUILDING   False     False   434       4.0  10.0   
997                       RESIDENCE   False     False  2011      20.0  40.0   
998                          STREET   False     False   331       3.0   NaN   
999                RESIDENCE-GARAGE   False     False   923       9.0  11.0   

     Community Area FBI Code  X Coordinate  Y Coordinate  Year  \
995            44.0       26     1178260.0     1846297.0  2014   
996            55.0       05     1191658.0     1814171.0  2017   
997             2.0       11     1154971.0     1938707.0  2020   
998             NaN       06     1192939.0     1858982.0  2001   
999            60.0       05     1170351.0     1884713.0  2003   

                 Updated On   Latitude  Longitude  \
995  02/04/2016 06:33:39 AM  41.733547 -87.622530   
996  02/10/2018 03:50:01 PM  41.645076 -87.574483   
997  04/08/2020 03:49:33 PM  41.987627 -87.705373   
998  08/17/2015 03:03:40 PM  41.768011 -87.568341   
999  02/10/2018 03:50:01 PM  41.839141 -87.650386   

                          Location  
995  (41.733547403, -87.622529593)  
996  (41.645075822, -87.574483491)  
997  (41.987626778, -87.705373183)  
998  (41.768010848, -87.568340708)  
999  (41.839140939, -87.650386262)

In [7]:
# Checking general information of the dataset
chicago_crime_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Unnamed: 0            1000 non-null   int64  
 1   ID                    1000 non-null   int64  
 2   Case Number           1000 non-null   object 
 3   Date                  1000 non-null   object 
 4   Block                 1000 non-null   object 
 5   IUCR                  1000 non-null   object 
 6   Primary Type          1000 non-null   object 
 7   Description           1000 non-null   object 
 8   Location Description  999 non-null    object 
 9   Arrest                1000 non-null   bool   
 10  Domestic              1000 non-null   bool   
 11  Beat                  1000 non-null   int64  
 12  District              1000 non-null   float64
 13  Ward                  929 non-null    float64
 14  Community Area        929 non-null    float64
 15  FBI Code              

In [8]:
# Checking statistical description of the dataset
chicago_crime_df.describe()

Unnamed: 0            ID         Beat     District        Ward  \
count   1000.000000  1.000000e+03  1000.000000  1000.000000  929.000000   
mean   53499.500000  6.936368e+06  1202.313000    11.507000   23.571582   
std      288.819436  3.362715e+06   711.664699     7.036154   14.293396   
min    53000.000000  2.260600e+04   111.000000     1.000000    1.000000   
25%    53249.750000  3.842856e+06   623.000000     6.000000   11.000000   
50%    53499.500000  7.032190e+06  1112.500000    11.000000   24.000000   
75%    53749.250000  9.846722e+06  1733.000000    17.000000   35.000000   
max    53999.000000  1.277162e+07  2535.000000    25.000000   50.000000   

       Community Area  X Coordinate  Y Coordinate         Year    Latitude  \
count      929.000000  9.870000e+02  9.870000e+02  1000.000000  987.000000   
mean        36.550054  1.164625e+06  1.887975e+06  2009.722000   41.848200   
std         21.866363  1.657587e+04  3.184801e+04     5.890961    0.087591   
min          1.000000  1.100317e+06  1.814171e+06  2001.000000   41.645076   
25%         22.000000  1.153314e+06  1.860908e+06  2005.000000   41.774051   
50%         32.000000  1.165973e+06  1.893913e+06  2009.000000   41.864496   
75%         55.000000  1.176432e+06  1.911064e+06  2014.000000   41.911833   
max         77.000000  1.204890e+06  1.951001e+06  2022.000000   42.021179   

        Longitude  
count  987.000000  
mean   -87.671349  
std      0.060334  
min    -87.906463  
25%    -87.712202  
50%    -87.665804  
75%    -87.628321  
max    -87.525488

In [9]:
# Checking number of rows and columns in our dataset
row, column = chicago_crime_df.shape

print(f'The number columns in the dataset is {column}.\nThe number of rows in the dataset is {row}.')

The number columns in the dataset is 23.
The number of rows in the dataset is 1000.


In [10]:
# Checking the columns in the dataset
print('The columns in this dataset are:')
for columns in chicago_crime_df.columns:
    print(columns)

The columns in this dataset are:
Unnamed: 0
ID
Case Number
Date
Block
IUCR
Primary Type
Description
Location Description
Arrest
Domestic
Beat
District
Ward
Community Area
FBI Code
X Coordinate
Y Coordinate
Year
Updated On
Latitude
Longitude
Location


In [11]:
# Checking the datatype for each of the columns in the dataframe
chicago_crime_df.dtypes

Unnamed: 0                int64
ID                        int64
Case Number              object
Date                     object
Block                    object
IUCR                     object
Primary Type             object
Description              object
Location Description     object
Arrest                     bool
Domestic                   bool
Beat                      int64
District                float64
Ward                    float64
Community Area          float64
FBI Code                 object
X Coordinate            float64
Y Coordinate            float64
Year                      int64
Updated On               object
Latitude                float64
Longitude               float64
Location                 object
dtype: object

#### 1. Use a funtion to Drop redundant columns (a redundant column is one that gives infromation that another column already explains: eg ID and Case number) 

In [12]:
def drop_columns(data, columns):
    """
    Drops specified columns from a DataFrame.

    Parameters:
    data (pd.DataFrame): The DataFrame from which columns are to be dropped.
    columns (list): The list of column names to be dropped.

    Returns:
    pd.DataFrame: The DataFrame after dropping the specified columns.
    """
    data.drop(columns=columns, inplace=True)
    return data


chicago_crime_df= drop_columns(chicago_crime_df, ['Unnamed: 0', 'X Coordinate', 'Y Coordinate', 'Latitude', 'Longitude', 'Location'])

chicago_crime_df.head()

ID Case Number                    Date                   Block  IUCR  \
0  4809324    HM300447  04/19/2006 04:45:02 PM      090XX S NORMAL AVE  1812   
1  2599849    HJ197748  02/21/2003 06:30:00 PM    039XX N AVONDALE AVE  0820   
2  7495181    HR183541  02/22/2009 07:42:00 PM  011XX S CALIFORNIA AVE  0486   
3  2628412    HJ231141  03/07/2003 02:46:00 PM      025XX W ADDISON ST  0890   
4  8389615    HT622350  12/06/2011 12:00:00 AM          072XX S MAY ST  2825   

    Primary Type                     Description      Location Description  \
0      NARCOTICS  POSS: CANNABIS MORE THAN 30GMS                 RESIDENCE   
1          THEFT                  $500 AND UNDER                    STREET   
2        BATTERY         DOMESTIC BATTERY SIMPLE                     OTHER   
3          THEFT                   FROM BUILDING  SCHOOL, PUBLIC, BUILDING   
4  OTHER OFFENSE         HARASSMENT BY TELEPHONE                 APARTMENT   

   Arrest  Domestic  Beat  District  Ward  Community Area FBI Code  Year  \
0   False     False  2223      22.0  21.0            71.0       18  2006   
1   False     False  1731      17.0  38.0            16.0       06  2003   
2   False     False  1135      11.0  28.0            29.0      08B  2009   
3   False     False  1913      19.0  47.0             5.0       06  2003   
4   False      True   733       7.0  17.0            68.0       26  2011   

               Updated On  
0  02/10/2018 03:50:01 PM  
1  02/28/2018 03:56:25 PM  
2  08/17/2015 03:03:40 PM  
3  02/28/2018 03:56:25 PM  
4  02/04/2016 06:33:39 AM

#### 2. Use functions to create new columns for Months, Day, Season.

In [13]:
# Converting the 'Date' column from an object type to a datetime type
chicago_crime_df['Date'] = pd.to_datetime(chicago_crime_df['Date'])

# Creating a function that extracts the day and month name from the 'Date' column
def day_month_column(data, column):
    '''
    Extracts and returns the day and month names from a date column in a DataFrame.

    Parameters:
    data (pd.DataFrame): The DataFrame containing the date column.
    column (str): The name of the date column from which to extract the day and month names.

    Returns:
    tuple: A tuple containing two pandas Series, the first with the day names and the second with the month names.
    '''
    day = data[column].dt.day_name()
    month = data[column].dt.month_name()

    return day, month
# Creating columns 'Day' and 'Month' using the 'day_month_column' function
chicago_crime_df['Day'], chicago_crime_df['Month'] = day_month_column(chicago_crime_df, 'Date')

# Creating function that creates instances for different seasons
def season(month):
    if month in [3, 4, 5]:
        return 'Spring'
    elif month in [6, 7, 8]:
        return 'Summer'
    elif month in [9, 10, 11]:
        return 'Autumn'
    else:
        return 'Winter'

# Creating column 'Season' using the 'season' function
chicago_crime_df['Season'] = chicago_crime_df.Date.dt.month.apply(season)

chicago_crime_df.head()

C:\Users\Alu\AppData\Local\Temp\ipykernel_1500\1422185103.py:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  chicago_crime_df['Date'] = pd.to_datetime(chicago_crime_df['Date'])


ID Case Number                Date                   Block  IUCR  \
0  4809324    HM300447 2006-04-19 16:45:02      090XX S NORMAL AVE  1812   
1  2599849    HJ197748 2003-02-21 18:30:00    039XX N AVONDALE AVE  0820   
2  7495181    HR183541 2009-02-22 19:42:00  011XX S CALIFORNIA AVE  0486   
3  2628412    HJ231141 2003-03-07 14:46:00      025XX W ADDISON ST  0890   
4  8389615    HT622350 2011-12-06 00:00:00          072XX S MAY ST  2825   

    Primary Type                     Description      Location Description  \
0      NARCOTICS  POSS: CANNABIS MORE THAN 30GMS                 RESIDENCE   
1          THEFT                  $500 AND UNDER                    STREET   
2        BATTERY         DOMESTIC BATTERY SIMPLE                     OTHER   
3          THEFT                   FROM BUILDING  SCHOOL, PUBLIC, BUILDING   
4  OTHER OFFENSE         HARASSMENT BY TELEPHONE                 APARTMENT   

   Arrest  Domestic  Beat  District  Ward  Community Area FBI Code  Year  \
0   False     False  2223      22.0  21.0            71.0       18  2006   
1   False     False  1731      17.0  38.0            16.0       06  2003   
2   False     False  1135      11.0  28.0            29.0      08B  2009   
3   False     False  1913      19.0  47.0             5.0       06  2003   
4   False      True   733       7.0  17.0            68.0       26  2011   

               Updated On        Day     Month  Season  
0  02/10/2018 03:50:01 PM  Wednesday     April  Spring  
1  02/28/2018 03:56:25 PM     Friday  February  Winter  
2  08/17/2015 03:03:40 PM     Sunday  February  Winter  
3  02/28/2018 03:56:25 PM     Friday     March  Spring  
4  02/04/2016 06:33:39 AM    Tuesday  December  Winter

#### 3. Use subsetting and grouping to show how the frequency of crime is distributed within Months, Days, Seasons with the most crime record and what crimes are highest and lowest?

In [26]:

# Showing the total number of crimes committed on each day in descending order and storing it in a variable 'crimes_each_day'
total_crimes_each_day = chicago_crime_df.groupby(['Day'])['Primary Type'].count().sort_values(ascending=False)
total_crimes_each_day

Day
Monday       169
Wednesday    147
Saturday     143
Sunday       139
Friday       135
Tuesday      134
Thursday     133
Name: Primary Type, dtype: int64

In [37]:
# Showing the total number of the types of crimes that happen each day
total_type_of_crimes_each_day = chicago_crime_df.groupby(['Day']).agg({'Primary Type' : ['value_counts']})
total_type_of_crimes_each_day

Primary Type
                                           value_counts
Day       Primary Type                                 
Friday    THEFT                                      33
          BATTERY                                    21
          DECEPTIVE PRACTICE                         11
          CRIMINAL DAMAGE                            11
          NARCOTICS                                  10
          OTHER OFFENSE                               9
          MOTOR VEHICLE THEFT                         9
          BURGLARY                                    6
          ROBBERY                                     6
          CRIMINAL TRESPASS                           5
          ASSAULT                                     4
          WEAPONS VIOLATION                           3
          PROSTITUTION                                2
          SEX OFFENSE                                 1
          STALKING                                    1
          KIDNAPPING                                  1
          HOMICIDE                                    1
          CRIM SEXUAL ASSAULT                         1
Monday    THEFT                                      42
          BATTERY                                    29
          CRIMINAL DAMAGE                            15
          BURGLARY                                   13
          NARCOTICS                                  13
          CRIMINAL TRESPASS                           9
          ASSAULT                                     9
          OTHER OFFENSE                               9
          MOTOR VEHICLE THEFT                         7
          ROBBERY                                     6
          DECEPTIVE PRACTICE                          6
          PROSTITUTION                                3
          PUBLIC PEACE VIOLATION                      2
          KIDNAPPING                                  1
          GAMBLING                                    1
          INTIMIDATION                                1
          OFFENSE INVOLVING CHILDREN                  1
          SEX OFFENSE                                 1
          WEAPONS VIOLATION                           1
Saturday  BATTERY                                    29
          THEFT                                      25
          CRIMINAL DAMAGE                            22
          NARCOTICS                                  12
          OTHER OFFENSE                               9
          BURGLARY                                    9
          ROBBERY                                     7
          ASSAULT                                     7
          MOTOR VEHICLE THEFT                         6
          CRIMINAL TRESPASS                           4
          DECEPTIVE PRACTICE                          4
          OFFENSE INVOLVING CHILDREN                  3
          INTERFERENCE WITH PUBLIC OFFICER            2
          SEX OFFENSE                                 2
          GAMBLING                                    1
          LIQUOR LAW VIOLATION                        1
Sunday    BATTERY                                    27
          THEFT                                      24
          CRIMINAL DAMAGE                            23
          NARCOTICS                                  16
          ASSAULT                                    10
          OTHER OFFENSE                               9
          ROBBERY                                     6
          BURGLARY                                    5
          MOTOR VEHICLE THEFT                         5
          DECEPTIVE PRACTICE                          5
          OFFENSE INVOLVING CHILDREN                  4
          CRIMINAL TRESPASS                           2
          PUBLIC PEACE VIOLATION                      1
          WEAPONS VIOLATION                           1
          HOMICIDE                                    1
Thursday  BATTERY                                    33
          T

In [29]:
# Showing the total number of crimes committed on each month in descending order and storing it in a variable 'crimes_each_day'
total_crimes_each_month = chicago_crime_df.groupby(['Month'])['Primary Type'].count().sort_values(ascending=False)
total_crimes_each_month

Month
July         96
April        94
October      93
March        90
May          87
September    87
January      79
August       77
December     76
June         76
November     76
February     69
Name: Primary Type, dtype: int64

In [38]:
# Showing the total number of the types of crimes that happen each month
total_type_of_crimes_each_month = chicago_crime_df.groupby(['Month']).agg({'Primary Type' : 'value_counts'})
total_type_of_crimes_each_month

Primary Type
Month     Primary Type                                  
April     THEFT                                       19
          BATTERY                                     16
          CRIMINAL DAMAGE                             10
          NARCOTICS                                    8
          OTHER OFFENSE                                8
          BURGLARY                                     7
          ASSAULT                                      6
          DECEPTIVE PRACTICE                           6
          CRIMINAL TRESPASS                            4
          MOTOR VEHICLE THEFT                          4
          ROBBERY                                      2
          WEAPONS VIOLATION                            2
          PUBLIC PEACE VIOLATION                       1
          PROSTITUTION                                 1
August    THEFT                                       16
          CRIMINAL DAMAGE                             14
          BATTERY                                     10
          OTHER OFFENSE                                9
          NARCOTICS                                    6
          ASSAULT                                      4
          MOTOR VEHICLE THEFT                          4
          ROBBERY                                      3
          OFFENSE INVOLVING CHILDREN                   2
          BURGLARY                                     2
          WEAPONS VIOLATION                            1
          SEX OFFENSE                                  1
          PUBLIC PEACE VIOLATION                       1
          GAMBLING                                     1
          DECEPTIVE PRACTICE                           1
          CRIMINAL TRESPASS                            1
          INTERFERENCE WITH PUBLIC OFFICER             1
December  THEFT                                       12
          BATTERY                                     11
          NARCOTICS                                   10
          CRIMINAL DAMAGE                              8
          BURGLARY                                     8
          ROBBERY                                      7
          OTHER OFFENSE                                5
          ASSAULT                                      3
          CRIMINAL TRESPASS                            3
          DECEPTIVE PRACTICE                           3
          MOTOR VEHICLE THEFT                          2
          INTERFERENCE WITH PUBLIC OFFICER             1
          INTIMIDATION                                 1
          OFFENSE INVOLVING CHILDREN                   1
          PUBLIC PEACE VIOLATION                       1
February  THEFT                                       16
          BATTERY                                     16
          CRIMINAL DAMAGE                              7
          OTHER OFFENSE                                6
          DECEPTIVE PRACTICE                           6
          ASSAULT                                      6
          NARCOTICS                                    5
          BURGLARY                                     3
          ROBBERY                                      2
          MOTOR VEHICLE THEFT                          1
          CRIMINAL TRESPASS                            1
January   THEFT                                       15
          BATTERY                                     13
          NARCOTICS                                   11
          CRIMINAL DAMAGE                              8
          MOTOR VEHICLE THEFT                          6
          DECEPTIVE PRACTICE                           5
          ROBBERY                                      5
          CRIMINAL TRESPASS                            4
          OTHER OFFENSE                                3
          BURGLARY                                     3
          ASSAULT                                      2
          WEAPONS VIOLATION                           

In [34]:
# Showing the total number of crimes committed on each season in descending order and storing it in a variable 'crimes_each_day'
total_crimes_each_season = chicago_crime_df.groupby(['Season'])['Primary Type'].count().sort_values(ascending=False)
total_crimes_each_season

Season
Spring    271
Autumn    256
Summer    249
Winter    224
Name: Primary Type, dtype: int64

In [39]:
# Showing the total number of the types of crimes that happen each season
total_type_of_crimes_each_season = chicago_crime_df.groupby(['Season']).agg({'Primary Type' : 'value_counts'})
total_type_of_crimes_each_season

Primary Type
Season Primary Type                                  
Autumn THEFT                                       65
       BATTERY                                     48
       CRIMINAL DAMAGE                             26
       NARCOTICS                                   20
       BURGLARY                                    19
       MOTOR VEHICLE THEFT                         15
       ASSAULT                                     14
       DECEPTIVE PRACTICE                          10
       ROBBERY                                      9
       OTHER OFFENSE                                9
       CRIMINAL TRESPASS                            6
       PROSTITUTION                                 5
       OFFENSE INVOLVING CHILDREN                   3
       SEX OFFENSE                                  2
       WEAPONS VIOLATION                            2
       INTIMIDATION                                 1
       GAMBLING                                     1
       LIQUOR LAW VIOLATION                         1
Spring THEFT                                       53
       BATTERY                                     52
       NARCOTICS                                   27
       CRIMINAL DAMAGE                             26
       OTHER OFFENSE                               23
       DECEPTIVE PRACTICE                          17
       BURGLARY                                    16
       ASSAULT                                     15
       MOTOR VEHICLE THEFT                         13
       ROBBERY                                     11
       CRIMINAL TRESPASS                           10
       WEAPONS VIOLATION                            2
       INTERFERENCE WITH PUBLIC OFFICER             2
       PROSTITUTION                                 1
       PUBLIC PEACE VIOLATION                       1
       KIDNAPPING                                   1
       OFFENSE INVOLVING CHILDREN                   1
Summer BATTERY                                     47
       THEFT                                       45
       CRIMINAL DAMAGE                             28
       OTHER OFFENSE                               23
       NARCOTICS                                   20
       ASSAULT                                     14
       DECEPTIVE PRACTICE                          13
       BURGLARY                                    13
       MOTOR VEHICLE THEFT                         11
       ROBBERY                                     10
       CRIMINAL TRESPASS                            7
       OFFENSE INVOLVING CHILDREN                   4
       PUBLIC PEACE VIOLATION                       2
       WEAPONS VIOLATION                            2
       SEX OFFENSE                                  2
       HOMICIDE                                     2
       PROSTITUTION                                 2
       CRIM SEXUAL ASSAULT                          1
       GAMBLING                                     1
       INTERFERENCE WITH PUBLIC OFFICER             1
       KIDNAPPING                                   1
Winter THEFT                                       43
       BATTERY                                     40
       NARCOTICS                                   26
       CRIMINAL DAMAGE                             23
       ROBBERY                                     14
       BURGLARY                                    14
       DECEPTIVE PRACTICE                          14
       OTHER OFFENSE                               14
       ASSAULT                                     11
       MOTOR VEHICLE THEFT                          9
       CRIMINAL TRESPASS                            8
       OFFENSE INVOLVING CHILDREN                   2
       WEAPONS VIOLATION                            2
       STALKING                                     1
       PUBLIC PEACE VIOLATION                       1
       INTERFERENCE WITH PUBLIC OFFICER             1
       INTIMIDATION                          

In [35]:
# Showing the total for each crime in the dataset and storing it in a variable 'frequency_of_crime'
frequency_of_crime = chicago_crime_df['Primary Type'].value_counts()
frequency_of_crime

Primary Type
THEFT                               206
BATTERY                             187
CRIMINAL DAMAGE                     103
NARCOTICS                            93
OTHER OFFENSE                        69
BURGLARY                             62
DECEPTIVE PRACTICE                   54
ASSAULT                              54
MOTOR VEHICLE THEFT                  48
ROBBERY                              44
CRIMINAL TRESPASS                    31
OFFENSE INVOLVING CHILDREN           10
PROSTITUTION                          8
WEAPONS VIOLATION                     8
PUBLIC PEACE VIOLATION                4
INTERFERENCE WITH PUBLIC OFFICER      4
SEX OFFENSE                           4
HOMICIDE                              2
INTIMIDATION                          2
GAMBLING                              2
KIDNAPPING                            2
LIQUOR LAW VIOLATION                  1
STALKING                              1
CRIM SEXUAL ASSAULT                   1
Name: count, dtype: int64

#### Conclusion on question 3:
From the first 2 cells we can see the frequency distribution of crimes in each day. The first cell shows that Monday is the day of the week with the highest crime committed, while the second cell shows that **theft** and **battery** are the crimes that is mosted committed each day of the week.

The third and fourth cells shows the frequency distribution of crimes in each month. The third cell shows that July is the month of the year with the highest crime committed, while the fourth cell shows that **theft** and **battery** are the crimes that is mosted committed each month of year.

The fifth and sixth cells shows the frequency distribution of crimes through each season. The fifth cell shows during spring season more crimes are committed, while the sixth cell shows that **theft** is the most committed crime across each season, closely followed by **battery** only during summer which the highest is **battery** and closely followed by **theft**. 

The seventh cell under question 3 shows the total for each crime committed across the whole data. It shows us that **theft** is the most committed crime with 206 occurrence in this data, followed by **battery** with 187 occurrence, while **Liquor Law Violation**, **Stalking**, and **Crim Sexual Assault** are the least committed crimes each with only 1 occurrence.

#### 4. According to location description, where does crime happen the most?

In [40]:
# Total number of crime committed on a specific location description in dataset
total_crime_committed_location = chicago_crime_df['Location Description'].value_counts()
total_crime_committed_location

Location Description
STREET                                            255
RESIDENCE                                         169
APARTMENT                                         123
SIDEWALK                                           87
OTHER                                              35
PARKING LOT/GARAGE(NON.RESID.)                     33
ALLEY                                              29
SMALL RETAIL STORE                                 24
RESTAURANT                                         24
RESIDENCE PORCH/HALLWAY                            17
RESIDENCE-GARAGE                                   15
VEHICLE NON-COMMERCIAL                             14
SCHOOL, PUBLIC, BUILDING                           14
GAS STATION                                        13
DEPARTMENT STORE                                   13
CHA PARKING LOT/GROUNDS                            11
GROCERY FOOD STORE                                  9
CTA PLATFORM                                        7
RESIDEN

In [41]:
# Showing the location with the highest crime committed in the dataset
print(f'The location with the highest crime rate is {total_crime_committed_location.idxmax()} with {total_crime_committed_location.max()} occurrence.')

The location with the highest crime rate is STREET with 255 occurrence.


In [47]:
chicago_crime_df.groupby(['Location Description'])['Primary Type'].count().sort_values(ascending=False)

Location Description
STREET                                            255
RESIDENCE                                         169
APARTMENT                                         123
SIDEWALK                                           87
OTHER                                              35
                                                 ... 
POOL ROOM                                           1
OTHER RAILROAD PROP / TRAIN DEPOT                   1
AIRPORT TERMINAL LOWER LEVEL - NON-SECURE AREA      1
HIGHWAY/EXPRESSWAY                                  1
ABANDONED BUILDING                                  1
Name: Primary Type, Length: 61, dtype: int64

#### 5. Did the types of crime change as years go by? If yes/no. Let the data show the insight.

In [42]:
# Showing the year for each crime commited and storing it in a variable 'crime_type_by_year'
crime_type_by_year = chicago_crime_df.groupby(['Primary Type', chicago_crime_df['Date'].dt.year])['Case Number'].count()
crime_type_by_year

Primary Type                      Date
ASSAULT                           2001     2
                                  2002     4
                                  2003     3
                                  2004     4
                                  2005     1
                                  2006     1
                                  2007     4
                                  2008     2
                                  2009     4
                                  2010     1
                                  2011     3
                                  2012     4
                                  2013     3
                                  2014     4
                                  2015     2
                                  2016     3
                                  2017     2
                                  2018     1
                                  2019     2
                                  2020     2
                                  2021     2
BATTERY         

In [51]:
chicago_crime_df[chicago_crime_df['Primary Type'] == 'ASSAULT'].head()

ID Case Number                Date                 Block  IUCR  \
12  11529170    JB545057 2018-11-27 09:47:00   004XX W HARRISON ST  0560   
13  11844612    JC452794 2019-09-29 06:40:00   071XX S ASHLAND AVE  051A   
17   5294640    HN154884 2007-02-01 12:45:39    079XX S DREXEL AVE  0560   
25   8690267    HV365694 2012-07-03 23:50:00    061XX S DREXEL AVE  051A   
26   2346895    HH646243 2002-09-13 20:56:00  005XX E BROWNING AVE  0530   

   Primary Type                    Description          Location Description  \
12      ASSAULT                         SIMPLE  GOVERNMENT BUILDING/PROPERTY   
13      ASSAULT            AGGRAVATED: HANDGUN                   GAS STATION   
17      ASSAULT                         SIMPLE                      SIDEWALK   
25      ASSAULT            AGGRAVATED: HANDGUN                      SIDEWALK   
26      ASSAULT  AGGRAVATED: OTHER DANG WEAPON                      SIDEWALK   

    Arrest  Domestic  Beat  District  Ward  Community Area FBI Code  \
12   False     False   124       1.0  25.0            28.0      08A   
13   False     False   734       7.0  17.0            67.0      04A   
17   False     False   624       6.0   8.0            44.0      08A   
25   False     False   313       3.0  20.0            42.0      04A   
26   False     False   212       2.0   4.0            35.0      04A   

    X Coordinate  Y Coordinate  Year              Updated On   Latitude  \
12     1173461.0     1897593.0  2018  12/12/2018 04:09:45 PM  41.874416   
13     1166911.0     1857552.0  2019  10/06/2019 04:07:52 PM  41.764682   
17     1183608.0     1852595.0  2007  02/10/2018 03:50:01 PM  41.750707   
25     1183066.0     1864341.0  2012  02/04/2016 06:33:39 AM  41.782952   
26     1180760.0     1881423.0  2002  02/28/2018 03:56:25 PM  41.829879   

    Longitude                       Location       Day      Month  Season  
12 -87.638592    (41.8744162, -87.638591667)   Tuesday   November  Autumn  
13 -87.663786  (41.764682386, -87.663785792)    Sunday  September  Autumn  
17 -87.602741  (41.750706929, -87.602741498)  Thursday   February  Winter  
25 -87.604363  (41.782951712, -87.604362828)   Tuesday       July  Summer  
26 -87.612292  (41.829879483, -87.612291851)    Friday  September  Autumn